### PR TITLE
Similarity learning reference code

### PR DESCRIPTION
**Pull request by dakshjotwani**
@fmassa PR for embedding learning reference code as discussed in #1042. I decided not to create a VGGFace2 dataset for now, since that would require more thought and planning. For now I'm using FMNIST.

### CODE REVIEW

File: references/embedding/loss.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    indices_not_equal = ~indices_equal
+
+    # Check if labels[i] == labels[j]
+    labels_equal = labels.unsqueeze(0) == labels.unsqueeze(1)
+
+    return labels_equal & indices_not_equal
+
+
+def _get_anchor_negative_triplet_mask(labels):
+    return ~(labels.unsqueeze(0) == labels.unsqueeze(1))
```
Comment (fmassa) [inline references/embedding/loss.py]: This is the same as `labels.unsqueeze(0) != labels.unsqueeze(1)`, but more efficient, right?

File: references/embedding/loss.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+'''
+    Pytorch adaptation of https://omoindrot.github.io/triplet-loss
+    https://github.com/omoindrot/tensorflow-triplet-loss
+'''
+import torch
+from torch.nn.modules.loss import _Loss
+from torch._jit_internal import weak_module, weak_script_method
+
+
+@weak_module
```
Comment (fmassa) [inline references/embedding/loss.py]: Can we remove the `weak_module` stuff? This is going away in PyTorch. You don't need to inherit from `_Loss`, which is an internal thing in PyTorch. Also, you don't use the `reduction` / `reduce` / `size_average` anywhere in your `forward`, so those can be all removed and just inherit from `nn.Module`.

File: references/embedding/sampler.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            keys_to_remove.append(key)
+            continue
+
+    for key in keys_to_remove:
+        group_samples.pop(key)
+
+    return group_samples
+
+
+class PKSampler(Sampler):
```
Comment (fmassa) [inline references/embedding/sampler.py]: Can you add some documentation to this sampler?

File: references/embedding/train.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            running_frac_pos_triplets = 0
+
+
+@torch.no_grad()
+def evaluate(model, dataset, device):
+    model.eval()
+    embeds, labels = None, None
+    dists, targets = None, None
+
+    loader = DataLoader(dataset, batch_size=512, shuffle=False, num_workers=4)
```
Comment (fmassa) [inline references/embedding/train.py]: Why not pass the dataloader as an argument to the function?
This keeps the `train` and `evaluate` functions consistent

File: references/embedding/train.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    for data in loader:
+        samples, _labels = data[0].to(device), data[1]
+        out = model(samples)
+        embeds = torch.cat((embeds, out), dim=0) if embeds is not None else out
```
Comment (fmassa) [inline references/embedding/train.py]: This is extremely inefficient. Can't you instead append to a list, and then perform a single `torch.cat` outside of the loop?

If the problem is memory consumption, you can then mode the embeddings to the CPU?

File: references/embedding/train.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    dists = torch.cdist(embeds, embeds)
+
+    labels = labels.unsqueeze(0)
+    targets = labels == labels.t()
+
+    mask = torch.ones(dists.size()).triu() - torch.eye(dists.size(0))
+    dists = dists[mask == 1]
+    targets = targets[mask == 1]
+
+    best_thresh = 0.01
```
Comment (fmassa) [inline references/embedding/train.py]: Can you factor this out into a helper function? This seems useful in general

File: references/embedding/loss.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    i_equal_k = label_equal.unsqueeze(1)
+
+    valid_labels = ~i_equal_k & i_equal_j
+
+    return valid_labels & distinct_indices
+
+
+def _get_anchor_positive_triplet_mask(labels):
+    # Check that i and j are distinct
+    indices_equal = torch.eye(labels.size(0)).byte().to(labels.device)
```
Comment (fmassa) [inline references/embedding/loss.py]: Can you instead do `torch.eye(labels.size(0), dtype=torch.uint8, device=labels.device)`? It's more efficient.

File: references/embedding/loss.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+    # Get final mean triplet loss over the positive valid triplets
+    triplet_loss = triplet_loss.sum() / (num_positive_triplets + 1e-16)
+
+    return triplet_loss, fraction_positive_triplets
+
+
+def _get_triplet_mask(labels):
+    # Check that i, j and k are distinct
+    indices_equal = torch.eye(labels.size(0)).byte().to(labels.device)
```
Comment (fmassa) [inline references/embedding/loss.py]: Same comment about passing `dtype` and `device` to `torch.eye`

File: references/embedding/model.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import torch
+import torch.nn as nn
+import torchvision.models as models
+
+
+class EmbeddingNet(nn.Module):
+    def __init__(self, backbone=models.resnet50(num_classes=128)):
```
Comment (fmassa) [inline references/embedding/model.py]: Can you make it `None` by default?
```python
def __init__(self, backbone=None):
    ...
    if backbone is None:
        backbone = models.resnet50(num_classes=128)
```

File: references/embedding/train.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
+import os
+
+import torch
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+from torch.utils.data.dataset import Subset
+from torch.nn.modules.distance import PairwiseDistance
```
Comment (fmassa) [inline references/embedding/train.py]: This doesn't seem to be used

File: references/embedding/train.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
+import os
+
+import torch
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+from torch.utils.data.dataset import Subset
```
Comment (fmassa) [inline references/embedding/train.py]: This isn't used it seems

File: references/embedding/train.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        save(model, epoch, args.save_dir, 'ckpt.pth')
+
+
+def parse_args():
+    import argparse
+    parser = argparse.ArgumentParser(description='PyTorch Embedding Learning')
+
+    parser.add_argument('--train-data', default='/tmp/pyemb/train/',
+                        help='FashionMNIST train dataset path')
+    parser.add_argument('--test-data', default='/tmp/pyemb/test/',
```
Comment (fmassa) [inline references/embedding/train.py]: At least for `FashionMNIST`, the root folder is the same, so this can probably be removed

File: references/embedding/train.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    optimizer = Adam(model.parameters(), lr=args.lr)
+
+    transform = transforms.Compose([transforms.Lambda(lambda image: image.convert('RGB')),
+                                    transforms.Resize((224, 224)),
+                                    transforms.ToTensor()])
+
+    train_dataset = FashionMNIST(args.train_data, train=True, transform=transform, download=True)
+    test_dataset = FashionMNIST(args.test_data, train=False, transform=transform, download=True)
+
+    targets = train_dataset.targets.tolist()
```
Comment (fmassa) [inline references/embedding/train.py]: A comment here would be helpful, as this is generally something that the user will need to change if they change the dataset

File: references/embedding/loss.py (hunk 1 above)
Comment (dakshjotwani) [reply references/embedding/loss.py]: I was thinking along the lines of `~(positive_triplet_mask)` when I wrote this. I agree this isn't as clean.

I just timed both these functions, and it turns out that `!=` is faster on average (Could be within margin of error). Should I change it to `!=`?

Comment (fmassa) [reply references/embedding/loss.py]: I didn't express myself properly in my first message, using `!=` is expected to be more efficient as it's a single operation, and not two as before. Can you change it to `!=`?

File: references/embedding/train.py (hunk 6 above)
Comment (dakshjotwani) [reply references/embedding/train.py]: Which part would you like to factor out? The upper triangular to vector section, or the embeddings to pairwise distance vector section?

File: references/embedding/loss.py (hunk 1 above)
Comment (dakshjotwani) [reply references/embedding/loss.py]: Yeah, done.

**Comment by dakshjotwani** [comment]
@fmassa I have made all the requested changes, other than the helper method change (I'm not sure which section you wanted to make a method).

File: references/embedding/loss.py (hunk 2 above)
Comment (dakshjotwani) [reply references/embedding/loss.py]: Done.

File: references/embedding/sampler.py (hunk 3 above)
Comment (dakshjotwani) [reply references/embedding/sampler.py]: I described what the sampler does and explained what `p` and `k` meant in the `Args:` section.

File: references/embedding/train.py (hunk 4 above)
Comment (dakshjotwani) [reply references/embedding/train.py]: I was going to do this, but I forgot about it before sending the PR. This has been changed.

File: references/embedding/train.py (hunk 5 above)
Comment (dakshjotwani) [reply references/embedding/train.py]: I tested this (cat after loop) approach (10000 test samples) on an 1080ti and 2080 and didn't encounter any memory problems.

I have made this change.

File: references/embedding/loss.py (hunk 7 above)
Comment (dakshjotwani) [reply references/embedding/loss.py]: Done

File: references/embedding/loss.py (hunk 8 above)
Comment (dakshjotwani) [reply references/embedding/loss.py]: Done.

File: references/embedding/model.py (hunk 9 above)
Comment (dakshjotwani) [reply references/embedding/model.py]: Done.

File: references/embedding/train.py (hunk 10 above)
Comment (dakshjotwani) [reply references/embedding/train.py]: Forgot to remove these before the PR. Sorry.

Removed.

File: references/embedding/train.py (hunk 12 above)
Comment (dakshjotwani) [reply references/embedding/train.py]: Changed the arg to `--dataset-dir`.

File: references/embedding/train.py (hunk 13 above)
Comment (dakshjotwani) [reply references/embedding/train.py]: I added comments mentioning that any classification dataset should be fine here as long as `targets` is constructed as described. Is that sufficient?

File: references/embedding/train.py (hunk 6 above)
Comment (fmassa) [reply references/embedding/train.py]: The accuracy computation. For example, by passing the `embeds` and the `labels`, returns the accuracy etc

File: references/embedding/train.py (hunk 13 above)
Comment (fmassa) [reply references/embedding/train.py]: The comment saying that it should be any classification dataset is misleading, because not all datasets have the `.targets` attribute, even if they are classification datasets. Maybe just check that the dataset has a `targets` attribute, and raise a nice error message if not?

File: references/embedding/sampler.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        p (int): Number of labels/groups to be sampled from in a batch
+        k (int): Number of samples for each label/group in a batch
+    """
+
+    def __init__(self, groups, p, k):
+        self.p = p
+        self.k = k
+        self.groups = create_groups(groups, self.k)
+
+    def __iter__(self):
```
Comment (fmassa) [inline references/embedding/sampler.py]: This is ok as is because we are not yet adding samplers to the main library, but once we move it to the `torchvision` package, it would be good to have tests for it.

If you could write a basic test now checking the behavior (with dummy data), it would make it much easier for moving this to torchvision core later on.

Comment (dakshjotwani) [reply references/embedding/sampler.py]: I have written the test cases for the sampler and refactored the accuracy section into another method.

I don't think we should assume that a targets attribute exists. Instead I feel like I can explain the targets data structure better, so that users can construct targets accordingly (or use the targets attribute if it exists). I have changed the comments slightly, explaining what is expected of the targets variable. Will this be okay instead?

**Comment by dakshjotwani** [comment]
@fmassa I have made the changes. Instead of expecting a targets attribute from the dataset, I elaborated further on the semantics and requirements from the targets data structure, which users can build during or after they have initialized their dataset. Will that be ok?

**Comment by dakshjotwani** [comment]
Renamed embedding to similarity to be more consistent with existing literature. Both are used, but similarity is more common.
